akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -196,13 +196,18 @@ private[stream] object Timers {
             pull(in)
           }
         }
+
+        override def onUpstreamFinish(): Unit = {
+          if (!isAvailable(in)) completeStage()
+        }
       })
 
       setHandler(out, new OutHandler {
         override def onPull(): Unit = {
           if (isAvailable(in)) {
             push(out, grab(in))
-            pull(in)
+            if (isClosed(in)) completeStage()
+            else pull(in)
           } else {
             if (nextDeadline.isOverdue()) {
```

Comment:
Depends on what do you consider a substate of what. I have no strong opinion on either. Maybe we need a tie-breaker. @viktorklang @ktoso ?